### PR TITLE
Add configuration resolution and utility functions

### PR DESCRIPTION
**Pull request by adamimos**
Added utilities to help resolve training configuration parameters and avoid redundant specification:

- config_resolution.py: Functions to compute generator sequence length and model vocab size from each other and special token usage
- persistence/utils.py: Checkpoint path parsing and step number formatting
- Device resolution for both JAX and PyTorch frameworks

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by adamimos** [comment]
@claude is this satisfactory now?

File: simplexity/persistence/utils.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
+import re
+
+
+def parse_checkpoint_step(path: str) -> int | None:
```
Comment (ealt) [inline simplexity/persistence/utils.py:4]: This assumes a certain set of naming conventions, and I think we should try to ensure that when code is coupled by shared standards/assumptions that is obvious in the codebase.

The convention `r"^(?P<dir>.+)/(?P<step>\d+)/model\.pt$"` comes from the [pytorch persister](https://github.com/Astera-org/simplexity/blob/bd938791572e98ab74353ec94f9b0789116a3c25/simplexity/persistence/local_pytorch_persister.py#L41), so either the parsing function should live in the same place as part of the persister or both functions should live together somewhere else.

I actually think maybe the best solution would be to write the `_get_path(self, step: int)` function here in the utils:

```python
def get_checkpoint_path(self, directory: Path, step: int, filename: str = "model.pt") -> Path:
    return directory / str(step) / filename
```

File: simplexity/persistence/utils.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    Examples:
+        >>> parse_checkpoint_step("model_weights/step_12345.pt")
+        12345
+        >>> parse_checkpoint_step("checkpoints/12345/model.pt")
+        12345
+        >>> parse_checkpoint_step("step-00500.pt")
+        500
+    """
+    m = re.search(r"step[_-]?(\d+)\.pt$", path)
```
Comment (ealt) [inline simplexity/persistence/utils.py:26]: This naming convention does not seem to be used in our codebase, so I would remove it

File: simplexity/persistence/utils.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    if m:
+        return int(m.group(1))
+
+    parts = path.split("/")
+    if parts and parts[-1] == "model.pt" and len(parts) >= 2:
```
Comment (ealt) [inline simplexity/persistence/utils.py:31]: If you use my suggested `get_checkpoint_path` function than the filename does not necessarily need to be `model.pt` (though maybe we want to make sure it is a valid filename with a `.pt` extension`)

File: simplexity/utils/config_resolution.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
+def compute_generator_sequence_length(model_n_ctx: int, use_bos: bool) -> int:
```
Comment (ealt) [inline simplexity/utils/config_resolution.py:1]: include `use_eos` as well

File: simplexity/utils/config_resolution.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    Examples:
+        >>> compute_generator_sequence_length(model_n_ctx=512, use_bos=True)
+        512
+        >>> compute_generator_sequence_length(model_n_ctx=512, use_bos=False)
+        513
+    """
+    return model_n_ctx + 1 - int(use_bos)
+
+
+def compute_model_context_length(generator_seq_len: int, use_bos: bool) -> int:
```
Comment (ealt) [inline simplexity/utils/config_resolution.py:24]: include `use_eos` as well

File: tests/persistence/test_utils.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        assert parse_checkpoint_step("step_00042.pt") == 42
+        assert parse_checkpoint_step("step_00000.pt") == 0
+        assert parse_checkpoint_step("0000/model.pt") == 0
+
+    def test_step_pattern_takes_precedence_over_directory(self):
+        """Test that step_*.pt pattern takes precedence over directory pattern."""
+        assert parse_checkpoint_step("path/step_200.pt") == 200
+        assert parse_checkpoint_step("checkpoints/step_999.pt") == 999
+
+    def test_windows_paths(self):
```
Comment (ealt) [inline tests/persistence/test_utils.py:75]: I don't understand how this test differs from previous ones, I also don't think we should worry about supporting Windows

File: tests/persistence/test_utils.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+        path_unix = "checkpoints/12345/model.pt"
+        assert parse_checkpoint_step(path_unix) == 12345
+
+    def test_s3_style_keys(self):
+        """Test S3 object key formats."""
+        assert parse_checkpoint_step("s3://bucket/prefix/step_12345.pt") == 12345
+        assert parse_checkpoint_step("prefix/run_name/12345/model.pt") == 12345
+
+
+class TestComputeStepWidth:
```
Comment (ealt) [inline tests/persistence/test_utils.py:86]: Don't need these tests

File: simplexity/persistence/utils.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    Examples:
+        >>> compute_step_width(999)
+        3
+        >>> compute_step_width(100000)
+        6
+    """
+    return len(str(max_steps))
+
+
+def format_step_number(step: int, max_steps: int) -> str:
```
Comment (ealt) [inline simplexity/persistence/utils.py:60]: This assumes `0 <= step <= max_steps`, so maybe we should explicitly make those assertions

**Comment by adamimos** [comment]
## Addressed all review comments

Thanks for the thorough review @ealt! I've addressed all your comments:

### Persistence Utils Changes:
1. ✅ **Inlined `compute_step_width()`** (line 40) - The function is now inlined directly into `format_step_number()`
2. ✅ **Added `get_checkpoint_path()` utility** (line 4) - Created a new function to centralize path construction and updated `local_pytorch_persister.py` to use it, making the coupling explicit
3. ✅ **Removed unused `step_*.pt` pattern** (line 26) - Removed the step_* and step- patterns since they're not used in the codebase
4. ✅ **Flexible filename support** (line 31) - The new `get_checkpoint_path()` accepts a `filename` parameter with `model.pt` as default
5. ✅ **Added input validation** (line 60) - Added explicit assertion `assert 0 <= step <= max_steps` to `format_step_number()`

### Config Resolution Changes:
6. ✅ **Added `use_eos` to `compute_generator_sequence_length()`** (line 1) - Full implementation with updated relationship formula
7. ✅ **Added `use_eos` to `compute_model_context_length()`** (line 24) - Full implementation with updated relationship formula

### Test Updates:
8. ✅ **Removed Windows path test** (line 75) - Removed redundant test
9. ✅ **Removed S3 tests** (line 86) - Removed unnecessary tests
10. ✅ **Removed `TestComputeStepWidth` class** - Function no longer exists
11. ✅ **Added `TestGetCheckpointPath` class** - Tests for the new utility function
12. ✅ **Updated config resolution tests** - Added comprehensive tests for `use_eos` parameter including round-trip consistency tests

All tests pass (100/100), ruff formatting/linting passes, and pyright type checking passes. Ready for re-review!

File: tests/persistence/test_utils.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+import pytest
+
+from simplexity.persistence.utils import format_step_number, get_checkpoint_path, parse_checkpoint_step
+
+
+class TestParseCheckpointStep:
+    """Test parse_checkpoint_step function."""
+
+    @pytest.mark.parametrize(
+        ("path", "expected"),
```
Comment (ealt) [inline tests/persistence/test_utils.py:12]: give some examples where the filename isn't `model.pt`

Comment (ealt) [reply tests/persistence/test_utils.py:12]: Also, give examples with zero padding

File: tests/persistence/test_utils.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+            "weights/model.eqx",
+            "random_file.txt",
+            "nonumeric/model.pt",
+        ],
+    )
+    def test_no_match_returns_none(self, path: str):
+        """Test paths that should not match any pattern."""
+        assert parse_checkpoint_step(path) is None
+
+    def test_zero_padded_step_numbers(self):
```
Comment (ealt) [inline tests/persistence/test_utils.py:39]: should be covered in previous test

File: tests/persistence/test_utils.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+
+    @pytest.mark.parametrize(
+        ("directory", "step", "filename", "expected"),
+        [
+            (Path("checkpoints"), 0, "model.pt", Path("checkpoints/0/model.pt")),
+            (Path("runs/exp1"), 1000, "checkpoint.pt", Path("runs/exp1/1000/checkpoint.pt")),
+            (Path("."), 42, "model.pt", Path("42/model.pt")),
+        ],
+    )
+    def test_parametrized_paths(self, directory: Path, step: int, filename: str, expected: Path):
```
Comment (ealt) [inline tests/persistence/test_utils.py:65]: does test anything meaningfully different from the tests above, remove

File: tests/persistence/test_utils.py [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+            (1, 999, "001"),
+            (42, 999, "042"),
+            (999, 999, "999"),
+            (0, 100000, "000000"),
+            (42, 100000, "000042"),
+            (12345, 100000, "012345"),
+            (100000, 100000, "100000"),
+        ],
+    )
+    def test_parametrized_formatting(self, step: int, max_steps: int, expected: str):
```
Comment (ealt) [inline tests/persistence/test_utils.py:101]: redundant with existing tests, remove

File: tests/persistence/test_utils.py [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+        """Test various step and max_steps combinations."""
+        assert format_step_number(step, max_steps) == expected
+
+    def test_lexicographic_ordering(self):
+        """Verify that formatted strings sort lexicographically."""
+        max_steps = 10000
+        formatted = [format_step_number(i, max_steps) for i in [1, 10, 100, 1000, 9999]]
+        assert formatted == sorted(formatted)
+
+    def test_width_computation(self):
```
Comment (ealt) [inline tests/persistence/test_utils.py:111]: not useful, remove

File: tests/persistence/test_utils.py [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+            (42, 100000, "000042"),
+            (12345, 100000, "012345"),
+            (100000, 100000, "100000"),
+        ],
+    )
+    def test_parametrized_formatting(self, step: int, max_steps: int, expected: str):
+        """Test various step and max_steps combinations."""
+        assert format_step_number(step, max_steps) == expected
+
+    def test_lexicographic_ordering(self):
```
Comment (ealt) [inline tests/persistence/test_utils.py:105]: please remove

File: tests/utils/test_config_resolution.py [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+        """When BOS is not used, generator_seq_len should be model_n_ctx + 1."""
+        assert compute_generator_sequence_length(model_n_ctx=512, use_bos=False) == 513
+        assert compute_generator_sequence_length(model_n_ctx=100, use_bos=False) == 101
+
+    def test_with_eos_token(self):
+        """When EOS is used, generator_seq_len should be model_n_ctx."""
+        assert compute_generator_sequence_length(model_n_ctx=512, use_bos=False, use_eos=True) == 512
+        assert compute_generator_sequence_length(model_n_ctx=100, use_bos=False, use_eos=True) == 100
+
+    def test_with_bos_and_eos(self):
```
Comment (ealt) [inline tests/utils/test_config_resolution.py:28]: consolidate these four tests with/without bos & eos into a single parameterized test with 4 test cases and the same model_n_ctx

File: tests/utils/test_config_resolution.py [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+
+    def test_with_bos_only(self):
+        """When only BOS is used, vocab size should be generator_vocab + 1."""
+        assert compute_model_vocab_size(generator_vocab_size=100, use_bos=True, use_eos=False) == 101
+
+    def test_with_eos_only(self):
+        """When only EOS is used, vocab size should be generator_vocab + 1."""
+        assert compute_model_vocab_size(generator_vocab_size=100, use_bos=False, use_eos=True) == 101
+
+    def test_with_both_special_tokens(self):
```
Comment (ealt) [inline tests/utils/test_config_resolution.py:148]: consolidate these four tests with/without bos & eos into a single parameterized test with 4 test cases and the same generator_vocab_size

File: tests/utils/test_config_resolution.py [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+            (100, False, True, 101),
+            (100, True, True, 102),
+            (1, False, False, 1),
+            (1, True, True, 3),
+            (50257, False, False, 50257),
+            (50257, True, False, 50258),
+            (50257, True, True, 50259),
+        ],
+    )
+    def test_parametrized_cases(self, generator_vocab_size: int, use_bos: bool, use_eos: bool, expected: int):
```
Comment (ealt) [inline tests/utils/test_config_resolution.py:166]: redundant, remove

File: tests/utils/test_config_resolution.py [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+        """Test various combinations of vocab size and special tokens."""
+        assert compute_model_vocab_size(generator_vocab_size, use_bos, use_eos) == expected
+
+    def test_minimal_vocab_with_tokens(self):
```
Comment (ealt) [inline tests/utils/test_config_resolution.py:170]: redundant, remove

File: tests/utils/test_config_resolution.py [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+        """Edge case: minimal vocabulary with special tokens."""
+        assert compute_model_vocab_size(generator_vocab_size=2, use_bos=True, use_eos=True) == 4
+
+    def test_large_vocab(self):
```
Comment (ealt) [inline tests/utils/test_config_resolution.py:174]: redundant, remove

File: tests/persistence/test_utils.py [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+            ("0/model.pt", 0),
+            ("prefix/run_name/12345/model.pt", 12345),
+        ],
+    )
+    def test_directory_model_format(self, path: str, expected: int):
+        """Test parsing {step}/model.pt format."""
+        assert parse_checkpoint_step(path) == expected
+
+    @pytest.mark.parametrize(
+        "path",
```
[the remark this answers is not in the document]
Comment (ealt) [reply tests/persistence/test_utils.py:26]: also, try to keep the number of test cases to a minimum

File: tests/persistence/test_utils.py (hunk 9 above)
Comment (ealt) [reply tests/persistence/test_utils.py:12]: also, try to keep the number of test cases to a minimum

File: tests/utils/test_jnp.py [hunk 21]
```diff
@@ -43,3 +44,85 @@ def test_signed_logsumexp():
...
+
+        with pytest.raises(RuntimeError, match="GPU requested but no GPU devices available"):
+            resolve_jax_device("cuda")
+
+    def test_invalid_spec_raises_value_error(self):
+        """Test invalid device spec raises ValueError."""
+        with pytest.raises(ValueError, match="Unknown device specification"):
+            resolve_jax_device("invalid_device")
+
+    def test_unknown_specs_raise_value_error(self):
```
Comment (ealt) [inline tests/utils/test_jnp.py:123]: redundant, remove

File: tests/utils/test_pytorch_utils.py [hunk 22]
```diff
@@ -40,3 +40,72 @@ def test_torch_to_jax(device: str):
...
+            pytest.skip("MPS is available, cannot test unavailable case")
+        with pytest.raises(RuntimeError, match="MPS requested but MPS is not available"):
+            resolve_device("mps")
+
+    def test_invalid_spec_raises_value_error(self):
+        """Test invalid device spec raises ValueError."""
+        with pytest.raises(ValueError, match="Unknown device specification"):
+            resolve_device("invalid_device")
+
+    def test_unknown_specs_raise_value_error(self):
```
Comment (ealt) [inline tests/utils/test_pytorch_utils.py:96]: redundant, remove

File: tests/utils/test_pytorch_utils.py [hunk 23]
```diff
@@ -40,3 +40,72 @@ def test_torch_to_jax(device: str):
...
+        """Test various unknown specs raise ValueError."""
+        invalid_specs = ["gpu", "cuda:0", "cuda:1", "tpu", "unknown"]
+        for spec in invalid_specs:
+            with pytest.raises(ValueError, match="Unknown device specification"):
+                resolve_device(spec)
+
+    def test_auto_mode_priority_order(self):
```
Comment (ealt) [inline tests/utils/test_pytorch_utils.py:103]: remove

File: simplexity/persistence/utils.py [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+        Step number if found, None otherwise
+
+    Examples:
+        >>> parse_checkpoint_step("checkpoints/12345/model.pt")
+        12345
+        >>> parse_checkpoint_step("12345/model.pt")
+        12345
+    """
+    parts = path.split("/")
+    if len(parts) >= 2 and parts[-1].endswith((".pt", ".eqx")):
```
Comment (ealt) [inline simplexity/persistence/utils.py:47]: Create a helper function that checks if a string is a valid filename with a supported extension to use both here as well as in `get_checkpoint_path` as an assertion that `filename` is valid

File: tests/persistence/test_utils.py (hunk 9 above)
Comment (ealt) [reply tests/persistence/test_utils.py:12]: I don't think we need 9 test cases, think about what important features each test case has and consolidate to a minimum set of test cases that covers all important features

File: tests/persistence/test_utils.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (ealt) [reply tests/persistence/test_utils.py:30]: I don't think we need 7 test cases, think about what important features each test case has and consolidate to a minimum set of test cases that covers all important features

File: simplexity/utils/config_resolution.py [hunk 25]
```diff
@@ -0,0 +1,99 @@
+def compute_generator_sequence_length(model_n_ctx: int, use_bos: bool, use_eos: bool = False) -> int:
```
Comment (ealt) [inline simplexity/utils/config_resolution.py:1]: give `use_bos` a default value of False and make both booleans keyword-only arguments

File: simplexity/utils/config_resolution.py [hunk 26]
```diff
@@ -0,0 +1,99 @@
...
+    result = model_n_ctx + 1 - int(use_bos) - int(use_eos)
+    if result <= 0:
+        raise ValueError(
+            f"Invalid configuration: model_n_ctx={model_n_ctx}, use_bos={use_bos}, use_eos={use_eos} "
+            f"results in non-positive generator sequence length ({result})"
+        )
+    return result
+
+
+def compute_model_context_length(generator_seq_len: int, use_bos: bool, use_eos: bool = False) -> int:
```
Comment (ealt) [inline simplexity/utils/config_resolution.py:38]: give `use_bos` a default value of False and make both booleans keyword-only arguments

File: simplexity/utils/config_resolution.py [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+    result = generator_seq_len - 1 + int(use_bos) + int(use_eos)
+    if result <= 0:
+        raise ValueError(
+            f"Invalid configuration: generator_seq_len={generator_seq_len}, use_bos={use_bos}, use_eos={use_eos} "
+            f"results in non-positive model context length ({result})"
+        )
+    return result
+
+
+def compute_model_vocab_size(generator_vocab_size: int, use_bos: bool, use_eos: bool = False) -> int:
```
Comment (ealt) [inline simplexity/utils/config_resolution.py:73]: give `use_bos` a default value of False and make both booleans keyword-only arguments

File: simplexity/persistence/utils.py [hunk 28]
```diff
@@ -0,0 +1,78 @@
+from pathlib import Path
+
+
+def get_checkpoint_path(directory: Path, step: int, filename: str = "model.pt") -> Path:
```
Comment (ealt) [inline simplexity/persistence/utils.py:4]: we should make use of `format_step_number` here so sorted checkpoint paths can naturally follow the ordering of the step number. Take an optional `max_steps` argument after filename and use `format_step_number` if the `max_steps` argument is supplied

File: tests/persistence/test_utils.py [hunk 29]
```diff
@@ -0,0 +1,101 @@
...
+    )
+    def test_no_match_returns_none(self, path: str):
+        """Test paths with numbers but invalid format return None."""
+        assert parse_checkpoint_step(path) is None
+
+
+class TestGetCheckpointPath:
+    """Test get_checkpoint_path function."""
+
+    @pytest.mark.parametrize(
```
Comment (ealt) [inline tests/persistence/test_utils.py:49]: Use an example filename with zero padding in the step number, and consolidate the total number of test cases

File: tests/persistence/test_utils.py (hunk 14 above)
Comment (ealt) [reply tests/persistence/test_utils.py:90]: please remove

File: tests/persistence/test_utils.py [hunk 30]
```diff
@@ -0,0 +1,101 @@
...
+    def test_negative_step_raises_error(self):
+        """Test that negative step values raise ValueError."""
+        with pytest.raises(ValueError, match="must be non-negative"):
+            get_checkpoint_path(Path("checkpoints"), -1)
+
+
+class TestFormatStepNumber:
+    """Test format_step_number function."""
+
+    @pytest.mark.parametrize(
```
Comment (ealt) [inline tests/persistence/test_utils.py:73]: too many test cases, please consolidate

File: tests/utils/test_config_resolution.py [hunk 31]
```diff
@@ -0,0 +1,156 @@
...
+            (512, True, False, 512),
+            (512, False, True, 512),
+            (512, True, True, 511),
+        ],
+    )
+    def test_bos_eos_combinations(self, model_n_ctx: int, use_bos: bool, use_eos: bool, expected: int):
+        """Test all combinations of BOS and EOS tokens with same model_n_ctx."""
+        assert compute_generator_sequence_length(model_n_ctx, use_bos, use_eos) == expected
+
+    @pytest.mark.parametrize(
```
Comment (ealt) [inline tests/utils/test_config_resolution.py:26]: too many test cases, reduce to 4 only keeping the ones with `model_n_ctx = 64`

File: tests/utils/test_config_resolution.py [hunk 32]
```diff
@@ -0,0 +1,156 @@
...
+
+
+class TestComputeModelContextLength:
+    """Test compute_model_context_length function."""
+
+    @pytest.mark.parametrize(
+        ("generator_seq_len", "use_bos", "use_eos", "expected"),
+        [
+            (513, False, False, 512),
+            (512, True, False, 512),
```
Comment (ealt) [inline tests/utils/test_config_resolution.py:61]: keep `generator_seq_len` fixed at 512 for all test cases and vary the `expected` values as appropriate

File: tests/utils/test_config_resolution.py [hunk 33]
```diff
@@ -0,0 +1,156 @@
...
+        """Test all combinations of BOS and EOS tokens with resulting model_n_ctx=512."""
+        assert compute_model_context_length(generator_seq_len, use_bos, use_eos) == expected
+
+    @pytest.mark.parametrize(
+        ("generator_seq_len", "use_bos", "use_eos", "expected"),
+        [
+            (1, True, False, 1),
+            (2, False, False, 1),
+            (1, False, True, 1),
+            (64, True, False, 64),
```
Comment (ealt) [inline tests/utils/test_config_resolution.py:76]: too many test cases, reduce to 4 with `generator_seq_len = 64` for each

File: tests/utils/test_config_resolution.py [hunk 34]
```diff
@@ -0,0 +1,176 @@
...
+        model_n_ctx = 512
+        use_bos = False
+        gen_seq_len = compute_generator_sequence_length(model_n_ctx, use_bos)
+        recovered_n_ctx = compute_model_context_length(gen_seq_len, use_bos)
+        assert recovered_n_ctx == model_n_ctx
+
+    @pytest.mark.parametrize("model_n_ctx", [1, 64, 128, 512, 1024])
+    @pytest.mark.parametrize("use_bos", [True, False])
+    @pytest.mark.parametrize("use_eos", [True, False])
+    def test_round_trip_consistency(self, model_n_ctx: int, use_bos: bool, use_eos: bool):
```
[the remark this answers is not in the document]
Comment (ealt) [reply tests/utils/test_config_resolution.py:114]: please remove

File: tests/utils/test_config_resolution.py [hunk 35]
```diff
@@ -0,0 +1,156 @@
...
+        ],
+    )
+    def test_bos_eos_combinations(self, generator_vocab_size: int, use_bos: bool, use_eos: bool, expected: int):
+        """Test all combinations of BOS and EOS tokens with same generator_vocab_size."""
+        assert compute_model_vocab_size(generator_vocab_size, use_bos, use_eos) == expected
+
+    @pytest.mark.parametrize(
+        ("generator_vocab_size", "use_bos", "use_eos", "expected"),
+        [
+            (1, False, False, 1),
```
Comment (ealt) [inline tests/utils/test_config_resolution.py:142]: remove test cases with `generator_vocab_size = 1`

**Comment by ealt** [comment]
@claude please address my comments
